File: JavaScript/algorithms/sort/merge_sort.js
const mergeSort = (nums) => {
  if (nums.length < 2) return nums;

  const midIndex = Math.floor(nums.length / 2);

  const left = nums.slice(0, midIndex);
  const right = nums.slice(midIndex);
  const sortedLeft = mergeSort(left);
  const sortedRight = mergeSort(right);
  return merge(sortedLeft, sortedRight);
};

const merge = (nums1, nums2) => {
  const times = nums1.length + nums2.length;
  nums1.push(Infinity);
  nums2.push(Infinity);

  const sorted = [];
  let i = 0;
  let j = 0;

  for (let k = 0; k < times; k++) {
    if (nums1[i] <= nums2[j]) {
      sorted.push(nums1[i]);
      i += 1;
    } else {
      sorted.push(nums2[j]);
      j += 1;
    }
  }

  return sorted;
};

console.log('Result', mergeSort([2, 23, 3, 31, 4, 1, 5, 7]));

File: Python/algorithms/sort_and_search/merge_sort.py
#! /usr/bin/python

import math

def mergeSort(nums):
    if len(nums) < 2:
        return nums

    mid = len(nums) // 2
    left = mergeSort(nums[:mid])
    right = mergeSort(nums[mid:])
    return merge(left, right)

def merge(left, right):
    result = []
    n = len(left) + len(right)
    left.append(math.inf)
    right.append(math.inf)

    l, r = 0, 0
    for i in range(n):
        if left[l] < right[r]:
            result.append(left[l])
            l += 1
        else:
            result.append(right[r])
            r += 1
    return result

if __name__ == "__main__":
    sample = [12, 5, 34, 6, 10, 11, 4, 2, 48, 32]
    print(mergeSort(sample))
